### DISCUSSION TITLE
Question: Rationale for V2 API?

### DISCUSSION THREAD

**Discussion by zaydek** [css-hooks/css-hooks#69] [General]
Hi there, I wanted to say thanks for CSS Hooks. I just picked it up a day or so ago and it's been exhilarating.

I'm wondering *why* the API changed in a meaningful way for V2 with respect to pseudos. Was the new $ API made to make CSS Hooks faster, or safer, or more reliable? I haven't found anything that goes into clear detail in the new docs, so I'd just like a better understanding of why the API changes were made and what problems they solve that perhaps the V1 didn't solve or in the least address.

Thank you, I'd be grateful for this clarity.

**Comment by zaydek** [comment]
Just found https://css-hooks.com/docs/configuration/#combinational-logic-hooks.

It looks like it has to do with composing nested selectors that can't be adjacent, such as media queries, which makes sense. I am wondering however if someone is _not_ intendeding to compose complex media queries if they can just use the simple syntax (V1) as a remedy?

**Comment by nsaunders** [comment]
Hi @zaydek, thanks for your interest in CSS Hooks, the kind words and the great question.

The main reason for the change is to allow the type of combinational logic you linked to be defined _locally_. Building on your example, let's say you have a few media query hooks `"@media (width < 500px)"`, `"@media (500px <= width < 1000px)"`, and `"@media (1000px <= width)"`. In v1, compositional nesting was strictly an _and_ operation, so the `4rem` margin below would never apply:

```typescript
css({
  "@media (500px <= width < 1000px)": {
    "@media (1000px <= width)": {
      margin: "4rem"
    }
  }
})
```

Clearly this is a case where we needed an _or_ condition (or at least a _not_ condition to invert `"@media (width < 500px)"`) but there's no way to express this using record syntax. (The logic can't be encoded in a record key.)

<details>
<summary><em>But wait! Are there other real-world use cases aside from media queries?</em></summary>

---

To be honest, I can't say for sure what people might do with locally-defined combinational logic in the real world. But [here](https://codesandbox.io/p/devbox/github/nsaunders/css-hooks-playground/tree/master/?file=%2Fsrc%2Fapp.tsx) is an early v2 prototype where I actually used (css) hooks to avoid component state changes and thus unnecessary re-renders. So I'm very interested to see where this goes, and if this technique can be used to make more components "stateless" from a React perspective.

---
</details>

Of course, if you've used the `Object.entries()` API much, you might already be familiar with the idea that a record is practically equivalent to an array of tuples - mostly just nicer syntax. So we can represent an _or_ condition this way:

```typescript
css({
  on: [
    [{ or: ["@media (500px <= width < 1000px)", "@media (1000px <= width)"] }, {
      margin: "4rem"
    }]
  ]
})
```

Notice that the `{ or: [...] }` bit is sort of like the "key", and `{ margin: "4rem" }` is the value. Definitely a bit more verbose, somewhat heavier syntax...but capable of representing _much_ more complex conditions.

Where the vanilla JSON syntax _really_ falls apart for me, however, is when applying Prettier:

```typescript
css({
  on: [
    [
      {
        or: [
          "@media (500px <= width < 1000px)",
          "@media (1000px <= width)",
        ],
      },
      {
        margin: "4rem",
      },
    ],
  ],
})
```

It looks like Prettier doesn't give any special treatment to tuples, so what _should_ look like a key/value as in a record (or CSS ruleset) ends up looking like a weird nested list structure.

That's where the `on` callback comes into play. Here's how it looks after applying Prettier:

```typescript
css({
  on: ($, { or }) => [
    $(or("@media (500px <= width < 1000px)", "@media (1000px <= width)"), {
      margin: "4rem",
    }),
  ],
})
```

This reads more like a CSS ruleset or even - if you really squint - a record.

As noted [here](https://css-hooks.com/docs/usage/#alternate-syntax), you can still use the plain JSON "list of tuples" syntax if you really need/want to, but I think most people will be happier with `$` the overwhelming majority of the time.

Aside from locally-defined combinational logic, I've also noticed some auxiliary benefits:
1. The "keys" (hook names / conditions) are separated into their own namespace (rather than being mixed with CSS property names), resulting a more concise list of editor hints.
2. Because it looks a bit less like CSS or "inline style Sass", I think some expectations around specificity and "deep merge" functionality are somewhat relaxed, which reduces confusion and helps me out from a maintainership perspective.

Finally, I do anticipate that the v2 API will be a bit "controversial" for early adopters, but I still think it's better to err on the side of more flexibility, composability, and expressive power, knowing that simplified abstractions can always be built on top. (On the other hand, you can't build flexibility, composability, and expressiveness on top of a basic/opinionated abstraction, a general design issue you might be familiar with.)

I hope this helps, and please let me know if you have any additional questions or thoughts on this.

Otherwise, thanks again, and I'm glad you're here!

**Comment by zaydek** [comment]
Thanks @nsaunders. I appreciate the thoughtfulness and completeness of your response. There are a few things I want to bring to your attention:

Don't media queries already have a way to express conditional logic and chaining, using `and` and `or`?

https://css-tricks.com/logic-in-css-media-queries/

For example:

```css
@media (min-width: 600px) and (max-width: 800px) {
  html { background: red; }
}
```

Here's an example that uses compound media queries (no Sass or transpilation used here)

https://codepen.io/zaydek/pen/ExMJdJO

https://github.com/css-hooks/css-hooks/assets/58870766/ce2f1826-437e-4929-8871-58fba004d614

Perhaps I'm missing something but this should demonstrate that in a single line of code you can express the compound relationship of _applying styles_ at >=1250px OR <1000px and >=500px, which if I'm not mistaken, is more or less what you are trying to accomplish with your `on/$` pattern.

That's why I'm a little confused because it seems like CSS can already solve for these use cases unless I'm misunderstanding something? Or perhaps you are targeting backwards compatibility?

My only preference is that whatever path CSS Hooks takes resembles what the future of CSS will look like anyway. Seeing as we recently got support for `&` nesting, and hopefully will get native pseudo support in inline styles in the future (fingers crossed), my ultimate hope is that in the future we can simply remove the `css` invokation, providing a real world way to write tomorrow's CSS today without significant migrations waiting for us on the other side (though some migration is inevitable, no matter what).

In any case, let's say you have another use case in mind where media query level 4 support wasn't as powerful or expressive or stateless as you intended, for everyone that doesn't have those pro user use cases, would it not make sense to provide them an API that allows them to write CSS the canonical way and then lean into your `on/$` pattern when they really need leverage, kind of like an escape hatch?

Ideally I can write (and perhaps I could in V1):

```js
css({
  color: "red",
  "@media (500px <= width < 1000px) or (width >= 1250px) {": {
    color: "blue",
  },
})
```

And there's no need for `on/$` and it's much easier to teach and maintain because no real learning needs to take place on behalf of the user.

---

As an aside, I can't tell you how happy CSS Hooks makes me. I am particularly excited about typed pseudo selectors. For example, I am currently experimenting with:

```js
import { createHooks } from "@css-hooks/react"

export const GROUP = "group"

const { styleSheet, css } = createHooks({
  hooks: {
    "&:hover":                      "&:hover",
    "&:hover:active":               "&:hover:active",
    "&:is(:hover, :focus-visible)": "&:is(:hover, :focus-visible)",

    [`.${GROUP}:is(:hover, :focus-visible) &`]: `.${GROUP}:is(:hover, :focus-visible) &`,
  },
  debug: import.meta.env.DEV,
})

export { css }

export function CSSHooksShim() {
  return <style dangerouslySetInnerHTML={{ __html: styleSheet() }} />
}
```

Perhaps in a future version, it would be possible to write inline pseudos without the need to instantiate the pseudos up front but I can appreciate there is probably some technical or performance aspect to this which requires pseudos be known ahead of time.

---

_Note on Prettier._

I've gone back and forth on using Prettier but for me it causes too much frustration than benefit, so I use a simple ESLint configuration that basically does 95% of the formatting I want without getting in the way. I combine that with VS Code's native support for formatting TypeScript and really I don't worry about formatting much anymore.

It's not as heavy handed as Prettier but I find it also allows me to write code a bit more like gofmt, which allows me to use newlines when I want to (it's entirely optional). So it's a bit more malleable than Prettier.

```js
// This is the React ESLint configuration file
module.exports = {
  "extends": [
    "plugin:react-hooks/recommended",
  ],
  "parser":  "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "rules":   {
    "arrow-parens":            ["warn", "always"],
    "comma-dangle":            ["warn", "always-multiline"],
    "indent":                  ["warn", "tab", { "SwitchCase": 1 }],
    "key-spacing":             ["warn", { "align": "value" }],
    "no-multiple-empty-lines": ["warn", { "max": 1 }],
    "quotes":                  ["warn", "double"],
    "semi":                    ["warn", "never"],
  },
}
```

**Comment by cly** [comment]
Having used tailwind and stylex recently, I think css-hooks beauty is in its simplicity and essentially being able to write inline styles. In stylex, one would compose these rules as per @zaydek (at the hook creation level). In tailwind, the media queries are pretty much fixed and rarely changed in most examples that I've seen. I too think v1's syntax is sufficient.

**Comment by zaydek** [comment]
Just a quick follow up, with respect to composing and React components:

I am trying to do something like the following:

```tsx
export function DropdownListItemTextAlt(props: JSX.IntrinsicElements["span"]) {
  return (
    <span
      {...props}
      style={css({
        opacity: 0.375,
        on:      ($) => [
          $(".group:is(:hover, :focus-visible) &", {
            opacity: 1,
            ...props.style, // Allow the consumer to override the styles (This doesn't work!)
          }),
        ],
        ...props.style, // Allow the consumer to override the styles
      })}
    />
  )
}

export function DropdownListItemText(props: JSX.IntrinsicElements["div"]) {
  return (
    <div
      {...props}
      style={css({
        ...truncate,
        color:      "white",
        fontFamily: "Inter, sans-serif",
        fontSize:   18/1.5,
        fontWeight: 400,
        ...props.style, // Allow the consumer to override the styles
      })}
    />
  )
}

export function DropdownListItem(props: JSX.IntrinsicElements["button"]) {
  return (
    <button
      {...props}
      className={group}
      style={css({
        alignItems:    "center",
        display:       "flex",
        gap:           8,
        paddingBlock:  3, // 18 + 3 + 3 = 24
        paddingInline: 16,
        on:            ($) => [
          $("&:hover", {
            backgroundColor: tokens.figma.blue,
          }),
        ],
        ...props.style, // Allow the consumer to override the styles
      })}
    />
  )
}
```

But I just realized with the V2 API I can't really imagine how to compose the variant styles. So ideally I would simply do:

```tsx
<DropdownListItemText>
  Hello,{" "}
  <DropdownListItemTextAlt>
    world
  </DropdownListItemTextAlt>
  !
</DropdownListItemText>
```

But if for some reason I wanted to override the group variant styles, such as:

```tsx
<DropdownListItemText>
  Hello,{" "}
  <DropdownListItemTextAlt
    style={css({
      color: "orange",
      ".group:is(:hover, :focus-visible) &": {
        color: "blue",
      },
    })}
  >
    world
  </DropdownListItemTextAlt>
  !
</DropdownListItemText>
```

I don't see how to do that because now I have to worry about callbacks, I'm not even sure it's possible.

So just to reiterate the main point -- by not having access to objects that can be predictable destructured so 'last wins', and even extend that to nested objects, for me it actually breaks composition from a user perspective where I can't easily reason about styling anymore. I know this is a bit of a niche use case but the point is to demonstrate that arbitrary composition is sort of broken with the current model, unless I'm doing something wrong.

What do you think? Am I putting too much reliance on CSS Hooks for this specific use case? Curious to hear your thoughts.

**Comment by nsaunders** [comment]
Hi @zaydek, I'll have more thoughts to share soon, but I wanted to offer a quick tip regarding style overrides to help unblock you:

The `css` function accepts multiple style object arguments (all of the same type). Each argument overrides the previous ones. So instead of this...

```tsx
export function DropdownListItem(props: JSX.IntrinsicElements["button"]) {
  return (
    <button
      {...props}
      className={group}
      style={css({
        alignItems: "center",
        display: "flex",
        gap: 8,
        paddingBlock: 3, // 18 + 3 + 3 = 24
        paddingInline: 16,
        on: ($) => [
          $("&:hover", {
            backgroundColor: tokens.figma.blue,
          }),
        ],
        ...props.style, // Allow the consumer to override the styles
      })}
    />
  );
}
```

...you'll want to do this...

```tsx
export function DropdownListItem(props: JSX.IntrinsicElements["button"]) {
  return (
    <button
      {...props}
      className={group}
      style={css(
        {
          alignItems: "center",
          display: "flex",
          gap: 8,
          paddingBlock: 3, // 18 + 3 + 3 = 24
          paddingInline: 16,
          on: ($) => [
            $("&:hover", {
              backgroundColor: tokens.figma.blue,
            }),
          ],
        },
        props.style, // Allow the consumer to override the styles
      )}
    />
  );
}
```

**Comment by robinweser** [comment]
I think the only downside with CSS media query `and` and `or` is that you would have to create hooks for each of them.
Apart from that I'm also, currently, leaning more towards the old API still as it's sufficient for me (I actually can't even think of a use case for `or` in media queries in any of my projects)
I think the only problem with being to similar to CSS is that nesting in CSS hooks doesn't actually work the way you would expect it in all cases. For example:
```js
const style = {
  "& img": {
    border: "1px solid black"
  }
}
```

You would assume that this adds a black border to all img children, right? This is not whats happening though. You cannot target and style other elements within your style object.

**Comment by nsaunders** [comment]
Regarding use cases, I agree that `and` and `not` are more often needed than `or`. However, there are undoubtedly _some_ use cases for `or`. In the CSS Hooks website itself, for example, I [used the `or` operator](https://github.com/css-hooks/css-hooks/blob/master/site/src/routes/docs/%5B...path%5D/layout.tsx#L209-L220) to conditionally show the nav menu as a function of both media queries and checkbox state. The alternatives might have been:

1. Centrally define a hook that would only be used once in this specific case; or
2. Implement a stateful checkbox that causes part of the UI to re-render when changed.

I agree that there is probably room for a "basic" abstraction (which looks more like v1) built on top of the new API. On the other hand, if that were the default, it wouldn't be possible to build a more "full-featured" API on top.

**Comment by cly** [comment]
In Tailwind, its possible to add new classes to the stylesheet by just using it. With css hooks and integrating with third party libraries such as react aria ive already had to add a number of specific hooks such as “&[data-pressed]” and then using it on the specific element. In tailwind it would be something like [data-pressed]:bg-red500.

It would be nice if css hooks can have a preprocessing step that simply takes all uses of hooks and puts them in the sheet without the need to manually define them via createhooks.

**Comment by nsaunders** [comment]
For those interested in a "basic" API resembling v1, please check out [css-hooks-basic](https://github.com/nsaunders/css-hooks-basic).

**Reply by zaydek** [reply]
Congrats that's really cool! I think that will help make it easier for people who want to adopt CSS Hooks and when they're ready to level up, they can lean into your canonical `and/or` API once the benefits outweigh the added syntax.

**Reply by mryechkin** [reply]
Thank you for this! The simplicity of the v1 API is what sold me on CSS Hooks originally, so appreciate you providing a way to still use it.

**Reply by nsaunders** [reply]
Certainly, I'm glad to have both options as well.

**Comment by nsaunders** [comment]
@zaydek, I wanted to share a quick note regarding the `&:is(:hover, :focus-visible)` hook.

It is perfectly valid, but there's a good chance you might end up having to define these hooks as well:

1. `&:hover`
2. `&:focus-visible`

I also noticed you have a `&:hover:active` hook. So now the list of centrally-defined hooks is growing...

1. `&:is(:hover, :focus-visible)`
2. `&:hover`
3. `&:focus-visible`
4. `&:hover:active`

... and you'll likely need to continue adding to this list as you run into new use cases.

It's not a horrible situation, but the ability to compose/invert hooks with arbitrary boolean logic can certainly improve it. You could just define these hooks instead:

1. `&:hover`
2. `&:focus-visible`
3. `&:active`

Then combine them as needed when styling an element:

```tsx
<button 
  style={css({
    on: ($, { and, or }) => [
      $(or("&:hover", "&:focus-visible"), {
        // &:is(:hover, :focus-visible) styles
      }),
      $(and("&:hover", "&:active"), {
        // &:hover:active styles
      })
    ]
  })}>
  Hello
</button>
```

This way, you've defined one fewer hook, and - more importantly - all of the hooks you have defined are general enough to be useful in future use cases - so you don't have to keep switching back to your `css.ts` to add more hooks.

The benefits are:
- reduced context switching
- reduced size of the hooks style sheet, which (assuming SSR) the client must download upfront before using your app

**Reply by zaydek** [reply]
Ah I agree, that is powerful. It's a way to destructure or desugar the compound selectors so you can compose that. That example you provided here perfectly clicked for me. I'd recommend using that example or something like it in the docs so to help articular the end user benefits of `and/or`.

**Comment by zaydek** [comment]
Thanks for the follow up -- I'll go ahead and close this since it seems like all the concerns have been fully addressed.
